rpc/compressor.go

```diff
@@ -11,18 +12,62 @@ type compressor interface {
 	Decompress([]byte) ([]byte, error)
 }
 
-type gzipCompressor struct{}
+type gzipCompressor struct {
+	writerLock  sync.Mutex
+	readerLock  sync.Mutex
+	gzipWriters []*gzip.Writer
+	gzipReaders []*gzip.Reader
+}
 
 var _ compressor = (*gzipCompressor)(nil)
 
 func newGzipCompressor() *gzipCompressor {
 	return &gzipCompressor{}
 }
 
+func (c *gzipCompressor) getGzipWriter(writer io.Writer) (*gzip.Writer, func()) {
+	c.writerLock.Lock()
+	defer c.writerLock.Unlock()
+	var gzipWriter *gzip.Writer
+	if len(c.gzipWriters) == 0 {
+		gzipWriter = gzip.NewWriter(writer)
+	} else {
+		gzipWriter = c.gzipWriters[0]
+		c.gzipWriters = c.gzipWriters[1:]
+		gzipWriter.Reset(writer)
+	}
+	return gzipWriter, func() {
+		c.writerLock.Lock()
+		c.gzipWriters = append(c.gzipWriters, gzipWriter)
+		c.writerLock.Unlock()
+	}
+}
+func (c *gzipCompressor) getGzipReader(reader io.Reader) (*gzip.Reader, func(), error) {
+	c.readerLock.Lock()
+	defer c.readerLock.Unlock()
+	var err error
+	var gzipReader *gzip.Reader
+	if len(c.gzipReaders) == 0 {
+		if gzipReader, err = gzip.NewReader(reader); err != nil {
+			return nil, func() {}, err
+		}
+	} else {
+		gzipReader = c.gzipReaders[0]
+		c.gzipReaders = c.gzipReaders[1:]
+		gzipReader.Reset(reader)
+	}
+	return gzipReader, func() {
+		c.readerLock.Lock()
+		c.gzipReaders = append(c.gzipReaders, gzipReader)
+		c.readerLock.Unlock()
+	}, nil
+}
+
 func (c *gzipCompressor) Compress(data []byte) ([]byte, error) {
 
 	var buf bytes.Buffer
-	writer := gzip.NewWriter(&buf)
+	writer, reclaim := c.getGzipWriter(&buf)
+	defer reclaim()
```

Comment:
I'm pretty sure writer.Close() will flush before it closes, so the writer.Flush() call is unnecessary.